src/content/components/TimelineViewport.js

```diff
@@ -0,0 +1,491 @@
+      (this: any)._mouseUpListener = this._mouseUpListener.bind(this);
+
+      (this: any)._setSize = this._setSize.bind(this);
+      (this: any)._setSizeNextFrame = this._setSizeNextFrame.bind(this);
+
+      this.shiftScrollId = 0;
+      this.zoomRangeSelectionScheduled = false;
+      this.zoomRangeSelectionScrollDelta = 0;
+
+      this.state = this.getDefaultState(props);
+    }
+
+    getHorizontalViewport({ selection, timeRange }: Props) {
+      if (selection.hasSelection) {
+        const { selectionStart, selectionEnd } = selection;
+        const timeRangeLength = timeRange.end - timeRange.start;
+        return {
+          viewportLeft: (selectionStart - timeRange.start) / timeRangeLength,
+          viewportRight: (selectionEnd - timeRange.start) / timeRangeLength,
+        };
+      }
+      return {
+        viewportLeft: 0,
+        viewportRight: 1,
+      };
+    }
+
+    getDefaultState(props: Props) {
+      const { viewportLeft, viewportRight } = this.getHorizontalViewport(props);
+      return {
+        containerWidth: 0,
+        containerHeight: 0,
+        containerLeft: 0,
+        viewportTop: 0,
+        viewportBottom: 0,
+        viewportLeft,
+        viewportRight,
+        dragX: 0,
+        dragY: 0,
+        isDragging: false,
+        isShiftScrollHintVisible: false,
+      };
+    }
+
+    /**
+     * Let the viewport know when we are actively scrolling.
+     */
+    showShiftScrollingHint() {
+      // Only show this message if we haven't shift zoomed yet.
+      if (this.props.hasZoomedViaMousewheel) {
+        return;
+      }
+
+      const scollId = ++this.shiftScrollId;
+      if (!this.state.isShiftScrollHintVisible) {
+        this.setState({ isShiftScrollHintVisible: true });
+      }
+      setTimeout(() => {
+        if (scollId === this.shiftScrollId) {
+          this.setState({ isShiftScrollHintVisible: false });
+        }
+      }, 1000);
+    }
+
+    componentDidUpdate(prevProps: Props) {
+      if (this.props.viewportNeedsUpdate(prevProps, this.props)) {
+        this.setState(this.getDefaultState(this.props));
+        requestAnimationFrame(this._setSize);
+      }
+    }
+
+    componentWillReceiveProps(newProps: Props) {
+      if (this.props.isRowExpanded !== newProps.isRowExpanded) {
+        this.setState(this.getDefaultState(newProps));
+        return;
+      }
+      if (
+        this.props.selection !== newProps.selection ||
+        this.props.timeRange !== newProps.timeRange
+      ) {
+        this.setState(this.getHorizontalViewport(newProps));
+      }
+    }
+
+    _setSize() {
+      const rect = this.refs.container.getBoundingClientRect();
+      if (this.state.containerWidth !== rect.width || this.state.containerHeight !== rect.height) {
+        const style = window.getComputedStyle(this.refs.container);
+
+        // Obey margins of the containing element.
+        const containerWidth = rect.width - parseFloat(style.marginLeft) - parseFloat(style.marginRight);
```

Comment:
I don't understand this. `rect` is the border box of this.refs.container. Why do you need to inset it by its own margin?